packages/ponybench/_auto_bench.pony

```diff
@@ -0,0 +1,98 @@
+use "collections"
+use "promises"
+
+actor _AutoBench[A: Any #share]
+  let _notify: _BenchNotify
+  let _run: {(U64)} val
+  let _auto_ops: _AutoOps
+
+  new create(
+    notify: _BenchNotify,
+    name: String,
+    f: ({(): A ?} val | {(): Promise[A] ?} val),
+    bench_time: U64 = 1_000_000_000,
+    max_ops: U64 = 100_000_000
+  ) =>
+    _notify = notify
+    _run = recover
+      lambda(ops: U64)(notify = this, name, f) =>
+        match f
+        | let fn: {(): A ?} val =>
+          _Bench[A](notify)(name, fn, ops)
+        | let fn: {(): Promise[A] ?} val =>
+          _BenchAsync[A](notify)(name, fn, ops)
+        end
+      end
+    end
+    _auto_ops = _AutoOps(bench_time, max_ops)
+
+  be apply(ops: U64 = 1) => _run(ops)
+
+  be _result(name: String, ops: U64, nspo: U64) =>
+    match _auto_ops(ops, ops*nspo, nspo)
+    | let ops': U64 => apply(ops')
+    else _notify._result(name, ops, nspo)
+    end
+
+  be _failure(name: String) => _notify._failure(name)
+
+class _AutoOps
+  let _bench_time: U64
+  let _max_ops: U64
+
+  new create(bench_time: U64, max_ops: U64) =>
+    (_bench_time, _max_ops) = (bench_time, max_ops)
+
+  fun apply(ops: U64, time: U64, nspo: U64): (U64 | None) =>
+    if (time < _bench_time) and (ops < _max_ops) then
+      var ops' = if nspo == 0 then
+        _max_ops
+      else
+        _bench_time / nspo
+      end
+      ops' = _max(_min(ops' + (ops' / 5), ops * 100), ops + 1)
+      _round_up(ops')
+    else
+      None
+    end
+
+  fun _min(x: U64, y: U64): U64 =>
+    if x > y then y else x end
+
+  fun _max(x: U64, y: U64): U64 =>
+    if x > y then x else y end
+
+  fun _round_down_10(x: U64): U64 =>
+    """
+    Round down to the nearest power of 10.
+    """
+    var tens: U64 = 0
+    // tens = floor(log_10(n))
+    var x' = x
+    while x' >= 10 do
+      x' = x' / 10
+      tens = tens + 1
+    end
+    // result = 10^tens
```

Comment:
Err... never mind - I see that `pow` is only defined for `F32` and `F64`, and not for integer types.